internal/services/rate.go

```diff
@@ -1,115 +1,160 @@
 package services
 
 import (
+	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/http"
-	"sync"
+	"strings"
 	"time"
-)
 
-const RatesAPIBaseURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+	"github.com/fdemchenko/exchanger/internal/cache"
+	"github.com/rs/zerolog/log"
+)
 
 const (
-	RetryInterval = 500 * time.Millisecond
-	RetryCount    = 3
+	NBUExchangeRateURL        = "https://bank.gov.ua/NBUStatService/v1/statdirectory/exchange?json"
+	FawazAhmedExchangeRateURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+
+	DefaultCachingDuration = 15 * time.Minute
+	RateFetchTimeout       = 10 * time.Second
 )
 
-type Rate struct {
+var ErrInvalidCurrencyCode = errors.New("invalid currency code")
+
+type ExchangeRateFetcher func(code string, client *http.Client) (float32, error)
+
+type NBUResponse struct {
+	Rate float32 `json:"rate"`
+	Code string  `json:"cc"`
+}
+
+func NBURateFetcher(code string, client *http.Client) (float32, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), RateFetchTimeout)
```

Comment:
What's the benefit of using `context.Background()` in such cases (in fawaz fetcher the same approach)? Why not to use request's context?

Reply:
I'm using `context.Background()` as placeholder, to add timeouts to requests. Indeed, passing request's context down the calls chain is the right way, and is the reason context package was created. I'll handle it.